scheds/rust/scx_lavd/src/bpf/main.bpf.c

```diff
@@ -1804,6 +1804,27 @@ void BPF_STRUCT_OPS(lavd_runnable, struct task_struct *p, u64 enq_flags)
 	waker_taskc->last_runnable_clk = now;
 }
 
+static bool need_to_calc_time_slice(struct task_struct *p)
+{
+	/*
+	 * We need to calculate the task @p's time slice in two cases: 1) if it
+	 * hasn't been calculated (i.e., LAVD_SLICE_UNDECIDED) after the
+	 * enqueue or 2) if the sched_ext kernel assigns the default time slice
+	 * (i.e., SCX_SLICE_DFL).
+	 *
+	 * Calculating and assigning a time slice without checking these two
+	 * conditions could entail pathological behaviors, notably watchdog
+	 * time out. One condition that could trigger a watchdog time-out is as
+	 * follows. 1) a task is preempted by another task, which runs in a
+	 * higher scheduler class (e.g., RT or DL). 2) when the task is
+	 * re-running after, for example, the RT task preempted out, its time
+	 * slice will be replenished again. 3) If these two steps are repeated,
+	 * the task can run forever.
+	 */
+	return p->scx.slice == LAVD_SLICE_UNDECIDED ||
+	       p->scx.slice == SCX_SLICE_DFL;
```

Comment:
So, the only time the kernel assigns `SCX_SLICE_DFL` is when the the currently running task is the only runnable task on the CPU. When the task's slice expires, the kernel sets slice to the default value and keeps running the task. This is a convenience feature which can be disabled by setting `SCX_OPS_ENQ_LAST` in `ops.flags`.  When the flag is set, the task will always be enqueued when the slice expires whether it's the last runnable task on the CPU or not. When the last task is enqueued, `ops.enqueue()` is called with `SCX_ENQ_LAST` flag:

```
        /*                                                                                                                                                                                                             
         * The task being enqueued is the only task available for the cpu. By                                                                                                                                          
         * default, ext core keeps executing such tasks but when                                                                                                                                                       
         * %SCX_OPS_ENQ_LAST is specified, they're ops.enqueue()'d with the                                                                                                                                            
         * %SCX_ENQ_LAST flag set.                                                                                                                                                                                     
         *                                                                                                                                                                                                             
         * If the BPF scheduler wants to continue executing the task,                                                                                                                                                  
         * ops.enqueue() should dispatch the task to %SCX_DSQ_LOCAL immediately.                                                                                                                                       
         * If the task gets queued on a different dsq or the BPF side, the BPF                                                                                                                                         
         * scheduler is responsible for triggering a follow-up scheduling event.                                                                                                                                       
         * Otherwise, Execution may stall.                                                                                                                                                                             
         */                                                                                                                                                                                                            
        SCX_ENQ_LAST            = 1LLU << 41,                                                                                                                                                                          
```

Reply:
For now, I will keep the code as it is but later when the preemption code per tick is ready, I will change it. 
Thank you!